src/pmdas/amdgpu/localdrm.c

```diff
@@ -0,0 +1,323 @@
+/*
+ * Copyright (c) 2024 Red Hat.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * for more details.
+ */
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#if defined(HAVE_DLFCN_H)
+#include <dlfcn.h>
+#endif
+#include "localdrm.h"
+
+#include <libdrm/amdgpu.h>
+#include <libdrm/amdgpu_drm.h>
+#include <xf86drm.h>
+
+#ifndef DSOSUFFIX
+#define DSOSUFFIX "so"
+#endif
+
+/* Looks like AMD kept PCI_VENDOR_ID_ATI for its GPU IDs */
+#define PCI_VENDOR_ID_ATI 0x1002
+
+int localDRMShutdown(drmDevicePtr devs[], uint32_t count) {
+    drmFreeDevices(devs, count);
+
+    free(devs);
+
+    return DRM_SUCCESS;
+}
+
+int localDRMDeviceGetDevices(drmDevicePtr *devs[],
+			     uint32_t *max,
+			     uint32_t *count) {
+  uint32_t amdgpu_count = 0;
+  drmDevicePtr *temp = NULL; /* Will contain all devices, including non-amd ones*/
+  drmDevicePtr *p = NULL; /* Helper to copy AMD device data */
+
+  /* First get the total count of devices */
+  int dev_count = drmGetDevices(NULL, 0);
+
+  if (dev_count <= 0) {
+      printf("No devices\n");
```

Comment:
The diagnostic printf calls in this file should all become __pmNotifyErr calls which interacts with the logging subsystem a bit more nicely (with timestamp prefixes, etc) - and ensures we write into the log and not stdout (which may even have pmcd on the other end of it waiting for a PDU).

Reply:
Oops, these are development artefacts, I'll rework/cleanup.